nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/ASURLClassLoader.java

```diff
@@ -887,11 +879,7 @@ private Vector<SentinelInputStream> getStreams() {
      * Must be synchnronized to (a) avoid race condition checking 'streams'.
      */
     private synchronized void closeOpenStreams() {
-        if (streams == null) {
-            return;
-        }
-        SentinelInputStream[] toClose = streams.toArray(new SentinelInputStream[streams.size()]);
-        for (SentinelInputStream s : toClose) {
+        for (SentinelInputStream s : streams) {
```

Comment:
I think we can get here `ConcurrentModificationException` unlike original version

Reply:
True.
Pffff, I opened the class in Eclipse now and it has much more issues ... 
* done() method should be a close() since 1.7. The closed classloader should not be used any more.
   * Such change may result in exceptions in incorrect implementations using this classloader, so we have to fix them.
* All those finalizers should be removed. Java [plans to remove them too](https://openjdk.org/jeps/421).
* Maybe we should continue step by step. I can't see all consequences yet.

Reply:
@dmatej Yep, I saw that too, but I wanted to change that step by step to avoid regression. Also finalizers have [been removed](https://github.com/payara/Payara/pull/6179) in Payara, so... I guess the same changes should be included in GlassFish.

Reply:
Fixed.